### ISSUE TITLE
Does "software-rendering" still have GPU requirements?

### ISSUE THREAD

**Issue by Xiaomita** [PCSX2/pcsx2#455]
Hello. With DX9 hardware we know that we need DX9 runtime support, a driver or exposure to the GPU that supports the DX9 implementation, and all relevant business surrounding this.

With software-rendering, however, it's all done with the CPU, but something still has to get those pixels to the screen somehow and there are several ways to do this, such as with Windows API, Cocoa, X Window.

Do you still need a DX9/GL2-compliant GPU in order for PCSX2 to use these libraries to put the pixels on the screen, or does PCSX2 use, say, just a rasterizer/bitblitting function via Windows API and needs no GPU at all?

Simply, can PCSX2 work with software-rendering WITHOUT a minimum DX9/GL2-compliant GPU?

By "work" I mean use the graphics plugins and run games, not just load the UI and window loop.

**Comment by pal1000** [comment]
Unfortunately, you need a really good GPU especially on OpenGL case even if you are using software renderer. See: #345 and https://communities.intel.com/thread/57145.

**Comment by Xiaomita** [comment]
Doesn't exactly answer my question though, but thanks.

**Comment by ssakash** [comment]
for some reasons even the GS values pumps up high while using software mode. so, I do think that a Good GPU is needed in case of software mode though, I'm not certain of the reasons.

**Comment by gregory38** [comment]
In sw rendering. Only a rendering of a texture quad is done. Technically it can be done with very basic function. However it is easier to do it with recent API and reuse code that is used in hw rendering.

What is the issue with dx9/gl3 ?

**Comment by gregory38** [comment]
Note are you aware that your *insert drawing api here* do use your GPU?

**Comment by ssakash** [comment]
yes, I'm aware of it but, how demanding is that on the gpu ?

**Comment by gregory38** [comment]
In the basic case (no interlace and extra effect)
For 1 frame:
* 1 single draw command
* 4 vertex shader runs
* your resolution ie "1920x1200" fragment shader runs

In short, it just paints a full screen once every frame for PC GPU is nothing. It even less expensive than to display modern desktop (transparency, multiple window, depth test...). Smartphone generally have a special unit to do this kind of operation.

In short you don't need a good GPU but you do need a good driver :)

**Comment by ssakash** [comment]
If, it's just a basic function Then, why is the load on gpu very high while using software mode ? Most of the time , the gs values go around 75-90% while playing games like ratchet and clank in software mode.

**Comment by gregory38** [comment]
The % is a measurement of CPU time. If you want to measure GPU activities you need to measure it with an external tool. Just check memory/core frequency, I'm nearly sure the GPU stay in low power.

**Comment by ssakash** [comment]
Yeah, you are right. the gpu usage is lower while observing through gpu-z in software mode. Damn,I was thinking that the gs values were the amount of the gpu's usage on the GS. Care to explain how it actually works ?

**Comment by avih** [comment]
@ssakash, this is off topic. Please ask at the forum.

**Comment by gregory38** [comment]
There is a function to measure the time spend in a thread. Then you divide it by the time spent to render a PS2 frame (between 2 PS2 vsync).

But, yeah that clearly off topic.

@Xiaomita are the various info/anwser OK for you? Otherwise please explain exactly what you want.

**Comment by ssakash** [comment]
Okay, I wil stop of off-topic questions here. Coming to the initial question by xiaomita , can pcsx2 work with software rendering without a minimum GL-2/DX-9 compliant gpu ?

**Comment by Bigpet** [comment]
The answer is already made, it currently does not work without a GPU API. But you could easily replace the full-screen quad with a blit, if you needed for some reason, but no one presented a good reason.

**Comment by avih** [comment]
There is a good reason to support simple SW renderer, such as running inside a linux vm for the sake of testing, or running under wine (e.g. on OS X or linux), where a simple SW renderer is often easier to get working than a backend which requires an actual GPU.

We've had SDL renderer in the past which did just that and worked fine under wine that, but it was removed around r57xx IIRC. (note that wine is also able to use some HW DX backends, but it's still more likely to fail than a plain SW backend).

However, this is a known issue, and if someone wants to help with that, patches are welcome.

**Comment by Xiaomita** [comment]
So what I'm getting is this:

Software renderer DOES use API functions in DX/GL still; it just does the heavy-lifting with the CPU and only uses the GPU-compliant functions because HW rendering uses them so the devs just continued with the same API; however, Windows API without DX/GL and without Aero, at least on Windows 2000/7, does NOT need hardware-acceleration, and can display the desktop/API programs without any DX/GL GPU stuff, so long as you only limit yourself to the API function built into Windows and do not use WGL/DX/other extension/wrappers; it just uses, I believe, a VGA driver that points to video MMIO ranges to manipulate pixels.

BUT ... software-rendering COULD only just use the Windows API functions and avoid DX/GL altogether, I'm assuming, but this would be extremely poor and lacking purpose if it did. The point of SW is to get the CPU to do most of the work rather than HW, which uses both CPU and GPU more elegantly.

If I got it right, let me know and this issue could happily be closed by me, the issuer, if desired.

**Comment by gregory38** [comment]
Even win 2000 is likely to use hardware accelerated graphics (2D). If I remember correctly without GPU driver, resolutions was very limited. You still need the GPU do the scaling and filtering. PC without GPU is really the dark age (https://en.wikipedia.org/wiki/Graphics_processing_unit)

**Comment by avih** [comment]
Gregory, you keep ignoring the linux vm case and easy of setting up under wine, such as on OS X.

**Comment by gregory38** [comment]
No I don't ignore it. VM only requires an old opengl version (done by SDL by the way).

Here the code that need to be plugged into GSdx (only single face, no depth, no rotation):
http://nehe.gamedev.net/tutorial/texture_mapping/12038/

**Comment by avih** [comment]
I can test a patch if you have one ;)

**Comment by Bigpet** [comment]
@avih here I copied existing code together to show you how much slower the native Windows version is https://github.com/Bigpet/pcsx2/compare/blttest

Makes an extra unnecessary copy but doesn't convert RGB->BGR, so that should be pretty much a zero-sum difference to doing it right
